File: 0115-distinct-subsequences/solution.cpp
class Solution {
public:   
    int numDistinct(string s, string t) {
        vector<vector<double>> dp(s.size() + 1, vector<double> (t.size() + 1, 0));
        vector<double> prev(t.size() + 1, 0);
        
        prev[0] = 1;
        
        for (int i = 1; i <= s.size(); ++i) {
            for (int j = t.size(); j >= 1; --j) {
                
                if (s[i - 1] == t[j - 1]) prev[j] = prev[j - 1] + prev[j];
                else prev[j] = prev[j];
                
            }
        }
        return (int)prev[t.size()];
    }
};

File: 0115-distinct-subsequences/solution.java
class Solution {
  public int numDistinct(String s, String t) {
    int m = s.length();
    int n = t.length();

    // 2D DP table where dp[i][j] stores the count of distinct subsequences
    int[][] dp = new int[m + 1][n + 1];

    // Base case: empty t can be formed 1 way (by deleting everything)
    for (int i = 0; i <= m; i++) {
      dp[i][0] = 1;
    }

    // Fill the DP table
    for (int i = 1; i <= m; i++) {
      for (int j = 1; j <= n; j++) {
        if (s.charAt(i - 1) == t.charAt(j - 1)) {
          // Include the match + exclude the match
          dp[i][j] = dp[i - 1][j - 1] + dp[i - 1][j];
        } else {
          // Characters don't match, carry over previous count from s
          dp[i][j] = dp[i - 1][j];
        }
      }
    }

    return dp[m][n];
  }
}
